staging/src/k8s.io/apimachinery/pkg/runtime/negotiate.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type clientNegotiator struct {
+	serializer     NegotiatedSerializer
+	encode, decode GroupVersioner
+}
+
+func (n *clientNegotiator) Encoder(contentType string, params map[string]string) (Encoder, error) {
```

Comment:
It might mean `pretty` is broken.  I'll double check.

Reply:
Pretty is handled in `NegotiateOutputMediaType`, none of the code that uses the client negotiator now previously handled Pretty.  I think that can be a follow up cleanup.

Reply:
Added a TODO

Reply:
it still feels broken to ignore the params... doesn't that mean it is really hard to use that to modify encoding/decoding in the future compatibly? Do we have scenarios that are populating these today? An initial implementation that returned an error if unknown params were required would make more sense to me (both for Encode/Decode/StreamDecoder)

Reply:
This is no different than the old code.  Client previously ignored these during negotiation.  In a sense these are informational - the only behavior we key off of today is pretty (at a higher level).

If we did add future behavior here, it would be backward compatible because this is client focused and already has to depend on the values in serializer info.  If they were populated by a server (hypothetically) we'd completely ignore them on the way back, and then a client that was smarter would have to start handling them.  If we started failing in the client now on unrecognized options, it's possible that we'd be broken against one or more server implementations (although very unlikely).

Reply:
> If we started failing in the client now on unrecognized options, it's possible that we'd be broken against one or more server implementations

Yeah, that's what I was wondering (like a LB that rewrote to `Content-Type: application/json;charset=utf8` or something supposedly no-op)